edx_exams/apps/api/v1/views.py

```diff
@@ -1 +1,164 @@
+from edx_api_doc_tools import schema
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+
+from edx_exams.apps.api.permissions import StaffUserPermissions
+from edx_exams.apps.api.serializers import ExamSerializer
+from edx_exams.apps.core.models import CourseExamConfiguration, Exam, ProctoringProvider
+
+
+log = logging.getLogger(__name__)
+
+
+class ExamsView(APIView):
+    """
+    View to modify exams.
+
+    HTTP PATCH
+    Creates a new Exam.
+    Expected PATCH data: {
+        "course_id": "courses-v1:edx+test+2019",
+        "content_id": 123,
+        "exam_name": "Test Examination",
+        "time_limit_mins": 30,
+        "due_date": "2021-08-04",
+        "exam_type": "proctored",
+        "hide_after_due": False,
+        "is_active": True
+    }
+    **PATCH data Parameters**
+        * course_id: The unique identifier for the course.
+        * content_id: This will be the pointer to the id of the piece of course_ware which is the proctored exam.
+        * exam_name: This is the display name of the Exam (Midterm etc).
+        * time_limit_mins: Time limit (in minutes) that a student can finish this exam.
+        * due_date: The date on which the exam is due
+        * exam_type: Type of exam, i.e. timed, proctored
+        * hide_after_due: Whether the exam will be hidden from the learner after the due date has passed.
+        * is_active: Whether this exam will be active.
+    **Exceptions**
+        * HTTP_400_BAD_REQUEST
+    """
+
+    authentication_classes = (JwtAuthentication,)
+    permission_classes = (StaffUserPermissions,)
+
+    @schema(
+      parameters=[],
+      responses={
+            200: "OK",
+            400: "Invalid request. See message."
+      },
+      summary='Modify exams',
+      description='This endpoint should create new exams, update existing exams, '
+                  'and mark any active exams not included in the payload as inactive.'
+    )
+    def patch(self, request, course_id):
+        """ create or update an exam """
+        request_exams = request.data
+
+        serializer = ExamSerializer(data=request_exams, many=True)
+
+        if serializer.is_valid():
+            course_exams = Exam.objects.filter(course_id=course_id)
+
+            visited_exams = []
+
+            for exam in request_exams:
+                exam_object = course_exams.filter(content_id=exam['content_id'])
+
+                # add to visited list so that we know to not mark this exam as inactive later
+                visited_exams.append(exam['content_id'])
+
+                if exam_object:
+                    if exam['exam_type'] != exam_object.exam_type:
+                        # if the exam type has changed, mark existing exam as inactive and then create
+                        # a new exam
+                        exam_object.is_active = False
+                        exam_object.save()
+
+                        current = Exam.objects.create(
```

Comment:
I could move some of this logic outside of the view to make this easier to read

Reply:
Is this a pattern you see repeated? If not, it's OK to see this here.

Reply:
It would be easier to test this logic if the serializer were not in the loop so that there is a function just over the list of exams and I do think it's a lot of stuff that doesn't need to be in the view.

This could be:
view extracts exams and calls update_these_here_exams(exams)
update separates the list to call update_exam(exam) or create_exam(exam)

most of the interesting tests are in update_exam to handle the update and maybe retire old one

Reply:
😆 I meant the serializer was not "in the loop" in the sense of being involved in the most interesting function to test. What an unfortunate choice of phrase.